File: new/python/twosum_using_hashmap_two_pass.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # go through the list and create a hashmap withe val: index
        # if we have duplicates in the list, the hashmap will have the recent index visited while iterating through it
        # again go through the list and get the difference between target and current number in the iteration
        # if that difference is present in the hashmap and that number is not the current number in our list iteration
        # then we found the pair. if the difference and the current iteration's number are same (they are the same number having
        # same index), then that is not the pair, move to next val.
        hash_map = dict()
        for i, val in enumerate(nums):
            hash_map[val] = (
                i  # this hashmap will have the most recent index for that value
            )

        for i, val in enumerate(nums):
            difference = target - nums[i]
            if difference in hash_map and hash_map[difference] != i:
                return [
                    min(i, hash_map[difference]),
                    max(i, hash_map[difference]),
                ]  # we don't have to check min or max because hashmap will have recent index
        # and we are iterating from the beginning.

        return []
        # time: O(n)
        # space: O(n) because we could have no duplicates in nums and thus create a hashmap of n items

File: new/cpp/twosum_using_hashmap_two_pass.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // create a hashmap and put the value and index from the list of num into the hashmap so we don't have any duplicates. 
        // hashmap will be like this-> val : index. The index will be the most recent index visited in the list during iteration. 
        // go through the list again and find difference between the target and the current val, and see if that is present
        // in the hashmap, and also check if the diff is present it is not our current value, it is not the one with the same index
        // as current value, this is to check if we are seeing the current value and don't add current value twice to the output
        // if not present return empty list
        unordered_map<int, int> hashMap;
        for (int i = 0; i < nums.size(); i++) {
            hashMap[nums[i]] = i;
        }

        for (int i = 0; i < nums.size(); i++) {
            int difference = target - nums[i]; 
            if (hashMap.count(difference) && hashMap[difference] != i) { // hashMap.count(val) returns either 0 or 1 because there are no duplicates in map. 
                return {min(i, hashMap[difference]), max(i, hashMap[difference])};
            }
        }

        return {};
    }

    // time: O(n)
    // space: O(n) hashMap could contain all the values from nums b/c nums didn't have no duplicates
};
